Authn docs (#1151)
emqtt/emq-docs-cn: en_US/security/authn/authn.md

```hocon
authentication = [
  {
    mechanism = password_based
    password_hash_algorithm {
      name = plain
      salt_position = suffix
    }
    enable = true
    backend = redis
    cmd = "HMGET mqtt_user:${username} password_hash salt is_superuser"
    database = 1
    password = public
    redis_type = single
    server = "127.0.0.1:6379"
  },
  {
    mechanism = password_based
    backend = built_in_database
    user_id_type = username
    password_hash_algorithm {
        name = sha256
        salt_position = suffix
    }
    enable = false
  }
]

```

An authentication chain without any enabled authenticators allows anonymous access.

Authentication chains can be configured globally per protocol as well as per individual protocol listeners:

```hocon
# emqx.conf

# Global chain for MQTT protocol
authentication = [
  ...
]

listeners.quic.default {
  ...
  # Specific chain for `quic.default` MQTT listener
  authentication = [
    ...
  ]
}

gateway.stomp {
  ...

  # Global chain for STOMP protocol
  authentication = [
    ...
  ]

  listeners.tcp.default {
    ...
    # Specific chain for `tcp.default` STOMP listener
    authentication = [
      ...
    ]
  }
}

```

> Listener-specific chains completely override the global chains, i.e. if a listener has its own chain global chain

Reviewer: ```suggestion
Listener-specific chains completely override the global chains, i.e. if a listener has its own chain, then the global chain
```
?

* Authenticator extracts hashed password and salt from the database using provided queries/selectors.
* Hashes password provided by the client with the configured hashing algorithm and fetched salt.
* Securely compares the resulting hash with the hash extracted from the database.

The following password hashing algorithms are supported:

```hocon
# simple algorithms:
password_hash_algorithm {
  name = sha256             # plain, md5, sha, sha512
  salt_position = suffix    # prefix, disable
}

# bcrypt
password_hash_algorithm {
  name = bcrypt
}

# pbkdf2
password_hash_algorithm {
  name = pbkdf2
  mac_fun = sha256          # md4, md5, ripemd160, sha, sha224, sha384, sha512
  iterations = 4096
  dk_length = 256           # optional
}
```

For password-based authenticators that allow user creation through EMQX API (`password_based:built_in_database`)
there are additional parameters required for hash creation:

```hocon
# bcrypt
password_hash_algorithm {
  name = bcrypt
  salt_rounds = 10          # used for user creation
}
```

## Authentication placeholders

Many authenticators allow using _placeholders_ in their configuration.

> Placeholders work as template varables that are filled with the corresponding

Reviewer: ```suggestion
Placeholders work as template variables that are filled with the corresponding
```

        "total": 0
      },
      "node": "emqx@127.0.0.1"
    }
  ],
  "node_resource_metrics": [
    {
      "metrics": {},
      "node": "emqx@127.0.0.1"
    }
  ],
  "node_status": [
    {
      "node": "emqx@127.0.0.1",
      "status": "connected"
    }
  ],
  "resource_metrics": {},
  "status": "connected"
}
```

#### POST /api/v5/listeners/{listener_id}/authentication/{id}/move

Move an authenticator within an MQTT listener authentication chain.

The following move commands are available:

* Move an authenticator to the beginning of the chain (it will be applied first during authentication):
  ```json
  {"position": "front"}
  ```
* Move to the end of the chain (it will be applied last during authentication):
  ```json
  {"position": "rear"}
  ```
* Place at the position preceding some other authenticator:
  ```json
  {"position": "before:password_based:built_in_database"}
  ```

```bash
## Request
curl -i \
--basic \
-u admin:public \
-X POST \
-H "Content-Type: application/json" \
http://localhost:18083/api/v5/listeners/tcp%3Adefault/authentication/password_based%3Abuilt_in_database/move \
-d @- <<'EOF'
{"position": "front"}
EOF

## Response
## 204 No Content
```

> ## TLS authentication

Reviewer 2: Move to `ssl.md`?

Author: ✅

Authentication in EMQX Broker means that when a client connects to EMQX Broker, the server configuration is used to control the client's permission to connect to the server.

EMQX Broker's authentication support includes two levels:

- The MQTT protocol itself specifies authentication primitives. EMQX Broker supports multiple variants of MQTT-level authentication:
  * username/password authentication with various backends (MongoDB, MySQL, PostgreSQL, Redis and built-in database);
  * SCRAM authentication with the built-in database;
  * JWT authentication;
  * authentication via custom HTTP API.
- At the transport layer, TLS guarantees client-to-server authentication using client certificates and ensures that the server verifies the server certificate to the client. PSK-based TLS/DTLS authentication is also supported.

In this document, we describe EMQX authentication and its configuration concepts.

## Authentication sources

_Authentication source_ (or simply _authenticator_) is an EMQX module that implements MQTT authentication. The following authenticators are
available by default:

> | mechanism        | backend            | description                                                                   |

Reviewer: A bit of clarification: _authenticators_ are defined just above, but here we have "mechanisms" and "backends".  Maybe point out how an authenticator maps to these?  For instance, an authenticator maps to a single (mechanism, backend) pair?

Author: ✅ 

Btw, now it seems for me that http authenticator shouldn't be `password_based` with `http` backend — there is no actual password verification (and http backend even may not check passwords).

Maybe, for http authenticator it would be more reasonable to have mechanism=`http` and do not have backend? 🤔

Reviewer: It sounds like HTTP is more akin to `jwt` authn.  But wouldn't `http` be more like the backend rather than the mechanism, since the latter can be something arbitrary for HTTP?  :thinking:

Author: M.b. something like `external:http`? With `external` for the mechanism and `http` for the backend?

Reviewer 3: Does this imply code/api changes?

Reviewer 3: Anyway, this can be a future improvement. Resolving the comment for now.

sequentially. Each authenticator may either return authentication success/failure or pass verification to
the next authenticator in the sequence. Such sequences are called _authentication chains_.

Conditions on which verification is passed further in the chain depend on the authenticator and described
in its documentation.

Each authentication chain can contain only one authenticator of each type.

Configured authentication chains can be updated or rearranged dynamically. Each authenticator
can be _enabled_ or _disabled_.

Example:

```hocon
authentication = [
  {
    mechanism = password_based
    password_hash_algorithm {
      name = plain
      salt_position = suffix
    }
    enable = true
    backend = redis
    cmd = "HMGET mqtt_user:${username} password_hash salt is_superuser"
    database = 1
    password = public
    redis_type = single
    server = "127.0.0.1:6379"
  },
  {
    mechanism = password_based
    backend = built_in_database
    user_id_type = username
    password_hash_algorithm {
        name = sha256
        salt_position = suffix
    }
    enable = false
  }
]

```

An authentication chain without any enabled authenticators allows anonymous access.

> Authentication chains can be configured globally per protocol as well as per individual protocol listeners:

Reviewer: Nit: maybe say whether the specific chains compose with or completely override the global chain.

Author: ✅

-u admin:public \
-X GET \
http://localhost:18083/api/v5/authentication/password_based%3Abuilt_in_database/status

## Response
{
  "metrics": {
    "failed": 0,
    "nomatch": 0,
    "rate": 0,
    "rate_last5m": 0,
    "rate_max": 0,
    "success": 0,
    "total": 0
  },
  "node_error": [],
  "node_metrics": [
    {
      "metrics": {
        "failed": 0,
        "nomatch": 0,
        "rate": 0,
        "rate_last5m": 0,
        "rate_max": 0,
        "success": 0,
        "total": 0
      },
      "node": "emqx@127.0.0.1"
    }
  ],
  "node_resource_metrics": [
    {
      "metrics": {},
      "node": "emqx@127.0.0.1"
    }
  ],
  "node_status": [
    {
      "node": "emqx@127.0.0.1",
      "status": "connected"
    }
  ],
  "resource_metrics": {},
  "status": "connected"
}
```

#### POST /api/v5/authentication/{id}/move

Move an authenticator within the global chain.

The following move commands are available:

* Move an authenticator to the beginning of the chain (it will be applied first during authentication):
  ```json
  {"position": "front"}
  ```
* Move to the end of the chain (it will be applied last during authentication):
  ```json
  {"position": "rear"}
  ```

> * Place to the position preceding some other authenticator:

Reviewer: ```suggestion
* Place at the position preceding some other authenticator:
```

## Response
{
  "metrics": {
    "failed": 0,
    "nomatch": 0,
    "rate": 0,
    "rate_last5m": 0,
    "rate_max": 0,
    "success": 0,
    "total": 0
  },
  "node_error": [],
  "node_metrics": [
    {
      "metrics": {
        "failed": 0,
        "nomatch": 0,
        "rate": 0,
        "rate_last5m": 0,
        "rate_max": 0,
        "success": 0,
        "total": 0
      },
      "node": "emqx@127.0.0.1"
    }
  ],
  "node_resource_metrics": [
    {
      "metrics": {},
      "node": "emqx@127.0.0.1"
    }
  ],
  "node_status": [
    {
      "node": "emqx@127.0.0.1",
      "status": "connected"
    }
  ],
  "resource_metrics": {},
  "status": "connected"
}
```

#### POST /api/v5/listeners/{listener_id}/authentication/{id}/move

Move an authenticator within an MQTT listener authentication chain.

The following move commands are available:

* Move an authenticator to the beginning of the chain (it will be applied first during authentication):
  ```json
  {"position": "front"}
  ```
* Move to the end of the chain (it will be applied last during authentication):
  ```json
  {"position": "rear"}
  ```

> * Place to the position preceding some other authenticator:

Reviewer: ```suggestion
* Place at the position preceding some other authenticator:
```

  ```json
  {"position": "before:password_based:built_in_database"}
  ```

```bash
## Request
curl -i \
--basic \
-u admin:public \
-X POST \
-H "Content-Type: application/json" \
http://localhost:18083/api/v5/listeners/tcp%3Adefault/authentication/password_based%3Abuilt_in_database/move \
-d @- <<'EOF'
{"position": "front"}
EOF

## Response
## 204 No Content
```

## TLS authentication

> To enable TLS authentication for clients, one may add an `ssl` listener

Reviewer: ```suggestion
To enable TLS authentication for clients, one may add a `ssl` listener
```

with `verify_peer` verify option set.

> The default `ssl` MQTT listener is run on the 8883 port:

Reviewer: ```suggestion
The default `ssl` MQTT listener runs on the 8883 port:
```
?

      salt_position = suffix
    }
    enable = true
    backend = redis
    cmd = "HMGET mqtt_user:${username} password_hash salt is_superuser"
    database = 1
    password = public
    redis_type = single
    server = "127.0.0.1:6379"
  },
  {
    mechanism = password_based
    backend = built_in_database
    user_id_type = username
    password_hash_algorithm {
        name = sha256
        salt_position = suffix
    }
    enable = false
  }
]

```

An authentication chain without any enabled authenticators allows anonymous access.

Authentication chains can be configured globally per protocol as well as per individual protocol listeners:

```hocon
# emqx.conf

# Global chain for MQTT protocol
authentication = [
  ...
]

listeners.quic.default {
  ...
  # Specific chain for `quic.default` MQTT listener
  authentication = [
    ...
  ]
}

gateway.stomp {
  ...

  # Global chain for STOMP protocol
  authentication = [
    ...
  ]

  listeners.tcp.default {
    ...
    # Specific chain for `tcp.default` STOMP listener
    authentication = [
      ...
    ]
  }
}

```

Listener-specific chains completely override the global chains, i.e. if a listener has its own chain global chain

> isn't used at all for clients connecting to the listener.

Reviewer 4: ```suggestion
Listener-specific chains completely override the global chains, i.e. if a listener has its own chain, the global chain
isn't used at all for the clients connecting to the listener.
```

> When a client connects to a listener it is authenticated with the listener-specific chain. If there is no
> chain specified for the listener, then the global chain for the listener protocol is used.

Reviewer 4: ```suggestion
When a client connects to a listener, it is authenticated with the listener-specific chain. If there is no
chain specified for the listener, the global chain for the listener protocol is used.
```

support multiple password hashing algorithms.

The algorithm for password verification is the following:
* Authenticator extracts hashed password and salt from the database using provided queries/selectors.
* Hashes password provided by the client with the configured hashing algorithm and fetched salt.
* Securely compares the resulting hash with the hash extracted from the database.

The following password hashing algorithms are supported:

```hocon
# simple algorithms:
password_hash_algorithm {
  name = sha256             # plain, md5, sha, sha512
  salt_position = suffix    # prefix, disable
}

# bcrypt
password_hash_algorithm {
  name = bcrypt
}

# pbkdf2
password_hash_algorithm {
  name = pbkdf2
  mac_fun = sha256          # md4, md5, ripemd160, sha, sha224, sha384, sha512
  iterations = 4096
  dk_length = 256           # optional
}
```

For password-based authenticators that allow user creation through EMQX API (`built_in_database`)
there are additional parameters required for hash creation:

```hocon
# bcrypt
password_hash_algorithm {
  name = bcrypt
  salt_rounds = 10          # used for user creation
}
```

> ## HTTP API

Reviewer 4: REST API?

Author: ✅

Authentication API allows to manipulate authentication chains and
concrete authenticators.

Authenticators are identified by their id formed as:
```
<mechanism>:<backend>
```
or just
```
<mechanism>
```
if there is no backend.

For example:

```
password_based:built_in_database
jwt
scram:built_in_database
```

::: tip
When used in URLs, authenticator ids should be URL-encoded: `password_based%3Abuilt_in_database`.
:::

### Global chain API

Global chain API operates with MQTT global (default) authentication chain.

#### GET /api/v5/authentication

> Get the global MQTT authentication chain.

Reviewer 4: ```suggestion
Get the global MQTT authentication chain:
```
?
Same for the rest of commands.

Author: This part was banished.

    {
      "metrics": {},
      "node": "emqx@127.0.0.1"
    }
  ],
  "node_status": [
    {
      "node": "emqx@127.0.0.1",
      "status": "connected"
    }
  ],
  "resource_metrics": {},
  "status": "connected"
}
```

#### POST /api/v5/listeners/{listener_id}/authentication/{id}/move

Move an authenticator within an MQTT listener authentication chain.

The following move commands are available:

* Move an authenticator to the beginning of the chain (it will be applied first during authentication):
  ```json
  {"position": "front"}
  ```
* Move to the end of the chain (it will be applied last during authentication):
  ```json
  {"position": "rear"}
  ```
* Place at the position preceding some other authenticator:
  ```json
  {"position": "before:password_based:built_in_database"}
  ```

```bash
## Request
curl -i \
--basic \
-u admin:public \
-X POST \
-H "Content-Type: application/json" \
http://localhost:18083/api/v5/listeners/tcp%3Adefault/authentication/password_based%3Abuilt_in_database/move \
-d @- <<'EOF'
{"position": "front"}
EOF

## Response
## 204 No Content
```

## TLS authentication

To enable TLS authentication for clients, one may add a `ssl` listener

> with `verify_peer` verify option set.

Reviewer 4: It's very hard to understand the meaning of this phrase. Did you mean

```
To enable TLS authentication for clients, one may set `ssl.verify` option to `verify_peer` for the listener.
```

Author: I tried to simplify it.
